### PR TITLE
Fix Codex tool-call poisoning and fatal KV completion races

### LINKED ISSUE #684: Recover conversations with malformed historical tool-call arguments
## Summary

Both `/v1/responses` and `/v1/chat/completions` reject an otherwise recoverable conversation when a historical assistant tool call contains non-empty malformed JSON arguments. This makes a missed or truncated tool call permanently poison all later retries that replay the conversation.

This is model- and client-independent frontend behavior. It was observed with Codex using the Responses API, but the failure is in shared chat-history preprocessing.

Related upstream report: https://github.com/vllm-project/vllm/issues/43995

## Reproduction

Send a Responses request containing this historical sequence:

```json
[
  {
    "type": "function_call",
    "call_id": "call_incomplete",
    "name": "diagnostic_noop",
    "arguments": "{\"value\": "
  },
  {
    "type": "function_call_output",
    "call_id": "call_incomplete",
    "output": "The tool was not executed because its arguments were invalid."
  },
  {
    "role": "user",
    "content": "Continue."
  }
]
```

The endpoint returns HTTP 400 with a JSON decode error while converting history for the chat template. A valid-argument history and a history without the malformed call both return HTTP 200.

## Root cause

`_postprocess_messages()` calls `json.loads()` unconditionally for every non-empty string in historical `tool_calls[].function.arguments`. A `JSONDecodeError` escapes before generation or chat-template rendering. Responses requests reach the same shared path after conversion to chat-completions-style history.

## Required behavior

- Preserve valid JSON behavior unchanged.
- Preserve malformed raw arguments as an explicit, valid, non-executable historical recovery record.
- Preserve tool-call IDs, matching tool outputs, and subsequent turns.
- Do not guess, repair, or execute malformed arguments.
- Do not log raw argument contents.
- Apply consistently to Responses and Chat Completions.

## Validation target

- Shared preprocessing unit coverage for valid, empty/null, and malformed arguments.
- Responses conversion coverage for malformed call + output + continuation.
- GLM chat-template render coverage showing the recovered history reaches the model safely.

### PR DESCRIPTION
**Pull request by yatesdr**
## Summary

This hardens the `dev/jovian-judgement` serving stack against the two production failures observed on GLM-5.3-Flash:

1. A streamed tool call can end with an incomplete JSON argument prefix. The parser now emits the shortest valid JSON suffix when normal finalization cannot extend the streamed prefix, while withholding unsafe separators and partial escapes.
2. If malformed arguments already exist in Responses or Chat Completions history, shared message preprocessing replaces them with `{}` and logs only the tool name and payload length. The malformed payload is never logged, reparsed by the chat template, or executed.
3. A late, duplicate, or out-of-order KV-transfer completion can no longer assert or delete an active request. Expected callbacks retain their existing behavior; stale callbacks are logged and ignored.

The target branch already contains the GLM delimiter-state fix from `9b38d88081`. This PR adds a regression using the captured GLM failure shape (a second `<tool_call>` beginning inside an unclosed argument) to verify the combined behavior.

Closes #684.

## Incident coverage

| Failure | Result after this PR |
|---|---|
| GLM emits/reopens tool markup before closing a streamed argument | Client receives parseable JSON containing the truncated/raw value, not a half-written JSON string |
| Codex replays a malformed `function_call.arguments` item | History is normalized before template rendering; later turns remain usable |
| LMCache engine-driven completion arrives after abort/cleanup | Scheduler ignores the unknown request instead of killing EngineCore |
| `finished_recving` and `finished_sending` arrive for the same aborted request in one step | Request is freed once; the second completion is treated as stale |
| Connector reports a completion for a still-active request | Callback is ignored and the active request is not freed |

## Upstream provenance and non-duplication

- Replay recovery is the adopted upstream behavior from vllm-project/vllm#48922 (merged). It is needed here because `dev/jovian-judgement` predates that merge.
- Streaming JSON finalization is a backport of vllm-project/vllm#53739, including its subsequent review fixes. That PR remains open upstream and its Qwen regression is parser-engine-wide; this PR adds the GLM production shape.
- Stale KV callback handling preserves the approach and authorship from vllm-project/vllm#37859, then extends it to validate request status and cover the same-step receive/send race reported in #46240. Related upstream attempts include #43265, #46304, #47499, and #49278; none is merged into this hardware branch.
- This supersedes local-inference-lab/vllm#685. That PR targets `main` and uses a local sentinel representation. This PR targets the branch used for these builds and follows the subsequently merged upstream `{}` behavior.

## Tests

Passed locally on macOS arm64 / Python 3.12 using an isolated `uv` environment:

```text
pytest tests/parser/engine/test_glm47_moe.py \
       tests/parser/engine/test_qwen3.py \
       tests/parser/engine/test_parser_engine.py -q
191 passed

pytest tests/entrypoints/unit_tests/test_chat_utils.py \
       tests/entrypoints/openai/responses/test_responses_utils.py \
       --confcutdir=tests/entrypoints \
       -k 'tool_call_arguments or malformed_function_call_history' -q
13 passed

VLLM_TARGET_DEVICE=cpu pytest tests/v1/core/test_scheduler.py \
       --confcutdir=tests/v1 \
       -k 'ignore_late_finished or same_step_recv_and_send_completion_after_abort or unexpected_kv_completion_does_not_free_active_request or abort_request_waiting_for_remote_kvs or abort_request_finished_recving or delayed_kv_connector_free_keeps_scheduler_active' -q
8 passed

pre-commit run --files <all 9 changed files>
All applicable checks passed, including ruff, formatting, typos, mypy, SPDX, forbidden imports, and configuration validation.
```

The `--confcutdir` options avoid vLLM's global macOS teardown fixture, whose `torch.accelerator.memory.empty_host_cache()` currently segfaults on this host; the tests themselves pass. Linux CI should run with the normal repository configuration.

Upstream #53739 also reports a forced-truncation live-model evaluation: 40/40 malformed tool calls before the parser fix and 0/40 after it.

## Model evaluation

No model-quality evaluation is applicable: this changes frontend history normalization, parser finalization, and scheduler lifecycle validation, not weights, kernels, sampling, or logits. The parser regression uses the captured GLM markup shape, and the scheduler regression reproduces the exact assertion ordering from the production report.

## AI assistance disclosure

OpenAI Codex was used for incident analysis, upstream research, implementation, and test execution. AI assistance is recorded in commit trailers. Upstream commits retain their original authorship. Per repository policy, the submitter must review every changed line before merge; normal human review is still required.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved handling of malformed, missing, or non-object tool-call arguments by normalizing them safely instead of raising errors.
  - Preserved tool-call history and identifiers when replaying conversations containing malformed arguments.
  - Improved streaming tool-call output so interrupted or incomplete values are completed as valid JSON while retaining completed data and embedded content.
  - Prevented stale or unexpected KV-transfer completions from crashing or incorrectly reviving requests.

### CODE REVIEW

**Comment by voipmonitor** [comment]
Validated the full PR in our composed GLM FP8 source, preserving the upstream and Derek's commit authorship. All changes apply cleanly, including the scheduler with atomic external-checkpoint import and compute-share fairness.

CPU results in the CUDA 13.3 / PyTorch 2.13 serving image:

- GLM/parser-engine suites: **137 passed**.
- Historical tool arguments and Responses replay: **13 passed**.
- Abort, stale/dual completion and active-request protection: **8 passed**.
- Installed-image GLM template rendering: **5 passed**. Truncated, non-object,
  empty and valid historical arguments render identically to their normalized
  mapping controls. Tool-call IDs, tool results and the following user turn
  remain intact. Only tokenizer assets were loaded; no model or tool ran.

No model, weight, sampling or GPU-kernel change is involved in these additions, so we are not adding another full throughput sweep for them. Qwen-specific tests were not run; no Qwen qualification is claimed.

One test-fixture observation: vLLM's default logger does not propagate to the root logger. Seven history assertions therefore see an empty plain `caplog.records` even though the expected warnings are emitted. Running the unchanged tests with `-o usefixtures=temporary_enable_log_propagate` passes all 13. Please use the repository's existing `caplog_vllm` fixture for those tests. The scheduler's CPU ExampleConnector fixture also requires unsetting the image's `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`; production LMCache supports that allocator independently.

The image composition also includes LMCache #61, which fixes the underlying dual-completion ownership defect reported in #698. Its unpatched abort matrix reproduced four failures; its complete patched adapter suite passes 67 tests. #685 is correctly superseded by this PR and is not being applied separately.

Validation used `/opt/venv/bin/python -m pytest` on the PR's selected test files.
The source-locked image is built; its installed GLM template probe passed with
target revision `46aaae8a82032f77100f2f03e9cc11b391df3b4d`. Its TP4/DCP1
DFlash2 API also returned final content exactly `12` for both valid and
malformed historical tool arguments, with the recorded tool result preserved.

Use the template's default reasoning mode for this API check. The GLM template
always appends `<think>` and does not implement `enable_thinking=false`.
Supplying that unsupported flag disables API reasoning extraction but leaves
the template unchanged, so even the valid-history control contains reasoning
in the content field. The failed fixture is retained separately; no serving
source change was needed. AI-assisted validation.

**Comment by voipmonitor** [comment]
Included in `dev/jovian-judgement` through [this PR's individual merge](https://github.com/local-inference-lab/vllm/commit/5198d889595b465a2963fb7e5448b3fc13d2ddce). The reviewed head and its contributor commits remain ancestors; the PR is merged and closed.

Source validation: replaying all 32 R35 review heads on the pinned base exactly reproduces the released Docker's vLLM tree; all 6,870 installed tracked files match. JJ additionally preserves Luke's DS4.1 work and #734. The final composition passed 247 focused checkpoint/scheduler, sampler/warmup and native GPU tests. This is combined-source evidence, not a fresh performance or full-model qualification for this individual PR.

Publication-history clarification: the individual merge linked above is in JJ's first-parent history. It replaces the receipt's archived wrapper-merge reference; GitHub's historical merge SHA may still identify that archive. See #731 for component review order and qualification limits.
